Added deep_transform_keys to StrongParameters

gustavogr opened the pull request:
### Summary

This PR implements the methods `deep_transform_keys` and `deep_transform_keys!` for `ActionController::Parameters`.

### Other Information

Thread in the Ruby On Rails Mailing list can be found [here](https://groups.google.com/forum/#!topic/rubyonrails-core/paiccJIC5E0).

gustavogr committed:
Implemented deep_transform_keys/! for ActionController::Parameters

eugeneius commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -695,6 +695,26 @@ def transform_keys!(&block)
+          @parameters.deep_transform_keys(&block)
+        )
+      else
+        @parameters.deep_transform_keys
```
It looks like this was copied from [the `transform_keys` implementation](https://github.com/rails/rails/blob/aa7da0471f7fcf2adbbc26c37a25914d8493928d/actionpack/lib/action_controller/metal/strong_parameters.rb#L687). It makes sense there, because calling `transform_keys` on a hash without passing a block returns an enumerator:

```ruby
irb(main):001:0> { foo: "bar" }.transform_keys
=> #<Enumerator: {:foo=>"bar"}:transform_keys>
```

However that doesn't work for `deep_transform_keys`:

```ruby
irb(main):002:0> { foo: "bar" }.deep_transform_keys
Traceback (most recent call last):
        7: from tools/console:11:in `<main>'
        6: from (irb):2
        5: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:66:in `deep_transform_keys'
        4: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:119:in `_deep_transform_keys_in_object'
        3: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:119:in `each_with_object'
        2: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:119:in `each'
        1: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:120:in `block in _deep_transform_keys_in_object'
LocalJumpError (no block given (yield))
```

Which means this method's implementation could be simplified:

```ruby
    def deep_transform_keys(&block)
      new_instance_with_inherited_permitted_status(
        @parameters.deep_transform_keys(&block)
      )
    end
```

GustavoTango replied:
Thank you for the feedback. I pushed a commit with your suggestion.

gustavogr replied:
Thank you for your feedback. Pushed a commit with your suggestion.

gustavogr commented:
Added a couple of tests in `accessors_test.rb` after looking at the changes in #36302. I rebased and squashed the changes in one commit following the contributors guide.
This is my first time contributing to Rails, so if there is more work to be done I'm happy to get to it.

eugeneius commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Added `deep_transform_keys` and `deep_transform_keys!` methods to StrongParameters.
```
StrongParameters looks like a module name here. Could you either change it to "strong parameters", or the actual class name (`ActionController::Parameters`)?

eugeneius commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -693,6 +693,22 @@ def transform_keys!(&block)
 
+    # Returns a new <tt>ActionController::Parameters</tt> instance with the
+    # results of running +block+ once for every key. This includes the keys
+    # from the root hash and from all nested hashes. The values are unchanged.
```
The documentation for `deep_transform_keys` mentions nested hashes **and arrays**:

https://github.com/rails/rails/blob/bbd065204ed45071f048badb7c0d49e35773d788/activesupport/lib/active_support/core_ext/hash/keys.rb#L58-L59

Could you include that here too? Obviously arrays don't have keys, but the point is that the method traverses nested arrays of hashes, not just nested hashes. If you can think of a clearer way to explain that, all the better 😊

gustavogr replied:
Maybe something like "_This method includes the keys of nested hashes and array of hashes_"?

eugeneius commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -693,6 +693,22 @@ def transform_keys!(&block)
+    end
+
+    # Returns the <tt>ActionController::Parameters</tt> instance changing its keys.
+    # This includes the keys from nested hashes.
```
Ditto on mentioning nested arrays here.

gustavogr commented:
Pushed new commit with changes in the documentation. Let me know if the documentation is still not clear.

eugeneius commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -693,6 +693,22 @@ def transform_keys!(&block)
 
+    # Returns a new <tt>ActionController::Parameters</tt> instance with the
+    # results of running +block+ once for every key. This method includes the
+    # keys of nested hashes and array of hashes. The values are unchanged.
```
The grammatically correct phrasing here would be "arrays of hashes".

I think I prefer the version used in the main `deep_transform_keys` docs though: "nested hashes and arrays of hashes" implies to me that only one level of nesting is covered, and the keys of a hash within an array-of-arrays wouldn't be transformed, for example, when in fact they would.

gustavogr commented:
Changed the documentation to match the version used in the main `deep_transform_keys` docs.

eugeneius commented:
This now looks good to me, thanks @gustavogr!

A committer will have to take a look before it can be merged.

gustavogr commented:
Thanks for your detailed help! @eugeneius
